src/state_causal_type.erl

```diff
@@ -0,0 +1,275 @@
+%% -------------------------------------------------------------------
+%%
+%% Copyright (c) 2015-2016 Christopher Meiklejohn.  All Rights Reserved.
+%% Copyright (c) 2007-2012 Basho Technologies, Inc.  All Rights Reserved.
+%%
+%% This file is provided to you under the Apache License,
+%% Version 2.0 (the "License"); you may not use this file
+%% except in compliance with the License.  You may obtain
+%% a copy of the License at
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @doc Common library for causal CRDTs.
+%%
+%% @reference Paulo Sérgio Almeida, Ali Shoker, and Carlos Baquero
+%% Delta State Replicated Data Types (2016)
+%% [http://arxiv.org/pdf/1603.01529v1.pdf]
+
+-module(state_causal_type).
+-author("Junghun Yoo <junghun.yoo@cs.ox.ac.uk>").
+
+%% causal_crdt() related.
+-export([new_causal_crdt/1,
+         causal_join/2,
+         is_lattice_inflation/2]).
+
+%% data_store() related.
+-export([new_data_store/1,
+         get_dot_cloud/1,
+         is_bottom_data_store/1]).
+%% dot_cloud() related.
+-export([get_next_dot_context/2,
+         insert_dot_context/2,
+         merge_dot_clouds/2]).
+%% {{dot_map, data_store_type()}, dot_map()} related.
+-export([get_sub_data_store/2,
+         insert_object/3,
+         remove_object/2,
+         get_all_objects/1,
+         get_objects_count/1]).
+
+-export_type([dot_context/0,
+              dot_cloud/0,
+              data_store/0,
+              data_store_type/0,
+              causal_crdt/0]).
+
+-type dot_actor() :: term().
+-type dot_context() :: {dot_actor(), pos_integer()}.
```

Comment:
This is a dot specifically, correct?  I just want to make sure my understanding is clear.